File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/FloatingLiteralTokenSyntax.cs
using Microsoft.VisualBasic.CompilerServices;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class FloatingLiteralTokenSyntax<T> : FloatingLiteralTokenSyntax
	{
		internal readonly T _value;

		internal T Value => _value;

		internal override string ValueText
		{
			get
			{
				T value = _value;
				return value.ToString();
			}
		}

		internal override object ObjectValue => Value;

		internal FloatingLiteralTokenSyntax(SyntaxKind kind, string text, GreenNode leadingTrivia, GreenNode trailingTrivia, TypeCharacter typeSuffix, T value)
			: base(kind, text, leadingTrivia, trailingTrivia, typeSuffix)
		{
			_value = value;
		}

		internal FloatingLiteralTokenSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, string text, GreenNode leadingTrivia, GreenNode trailingTrivia, TypeCharacter typeSuffix, T value)
			: base(kind, errors, annotations, text, leadingTrivia, trailingTrivia, typeSuffix)
		{
			_value = value;
		}

		internal FloatingLiteralTokenSyntax(ObjectReader reader)
			: base(reader)
		{
			_value = Microsoft.VisualBasic.CompilerServices.Conversions.ToGenericParameter<T>(reader.ReadValue());
		}

		static FloatingLiteralTokenSyntax()
		{
			ObjectBinder.RegisterTypeReader(typeof(FloatingLiteralTokenSyntax<T>), (ObjectReader r) => new FloatingLiteralTokenSyntax<T>(r));
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_value);
		}

		public override GreenNode WithLeadingTrivia(GreenNode trivia)
		{
			return new FloatingLiteralTokenSyntax<T>(base.Kind, GetDiagnostics(), GetAnnotations(), base.Text, trivia, GetTrailingTrivia(), _typeSuffix, _value);
		}

		public override GreenNode WithTrailingTrivia(GreenNode trivia)
		{
			return new FloatingLiteralTokenSyntax<T>(base.Kind, GetDiagnostics(), GetAnnotations(), base.Text, GetLeadingTrivia(), trivia, _typeSuffix, _value);
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new FloatingLiteralTokenSyntax<T>(base.Kind, newErrors, GetAnnotations(), base.Text, GetLeadingTrivia(), GetTrailingTrivia(), _typeSuffix, _value);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new FloatingLiteralTokenSyntax<T>(base.Kind, GetDiagnostics(), annotations, base.Text, GetLeadingTrivia(), GetTrailingTrivia(), _typeSuffix, _value);
		}
	}
	internal abstract class FloatingLiteralTokenSyntax : SyntaxToken
	{
		internal readonly TypeCharacter _typeSuffix;

		internal TypeCharacter TypeSuffix => _typeSuffix;

		internal FloatingLiteralTokenSyntax(SyntaxKind kind, string text, GreenNode leadingTrivia, GreenNode trailingTrivia, TypeCharacter typeSuffix)
			: base(kind, text, leadingTrivia, trailingTrivia)
		{
			_typeSuffix = typeSuffix;
		}

		internal FloatingLiteralTokenSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, string text, GreenNode leadingTrivia, GreenNode trailingTrivia, TypeCharacter typeSuffix)
			: base(kind, errors, annotations, text, leadingTrivia, trailingTrivia)
		{
			_typeSuffix = typeSuffix;
		}

		internal FloatingLiteralTokenSyntax(ObjectReader reader)
			: base(reader)
		{
			_typeSuffix = (TypeCharacter)reader.ReadByte();
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteByte((byte)_typeSuffix);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/FloatingLiteralTokenSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend MustInherit Class FloatingLiteralTokenSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SyntaxToken
		Friend ReadOnly _typeSuffix As TypeCharacter

		Friend ReadOnly Property TypeSuffix As TypeCharacter
			Get
				Return Me._typeSuffix
			End Get
		End Property

		Friend Sub New(ByVal kind As SyntaxKind, ByVal text As String, ByVal leadingTrivia As GreenNode, ByVal trailingTrivia As GreenNode, ByVal typeSuffix As TypeCharacter)
			MyBase.New(kind, text, leadingTrivia, trailingTrivia)
			Me._typeSuffix = typeSuffix
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal text As String, ByVal leadingTrivia As GreenNode, ByVal trailingTrivia As GreenNode, ByVal typeSuffix As TypeCharacter)
			MyBase.New(kind, errors, annotations, text, leadingTrivia, trailingTrivia)
			Me._typeSuffix = typeSuffix
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			Me._typeSuffix = reader.ReadByte()
		End Sub

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteByte(CByte(Me._typeSuffix))
		End Sub
	End Class
End Namespace